vdirsyncer/sync.py

```
@@ -99,107 +102,328 @@ def to_ident_conflict(self, storage):
+
+    @abc.abstractmethod
+    def update_ident_a(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def update_ident_b(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def remove_ident(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_a(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_b(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_new_a(self, ident):
+        raise NotImplementedError()
 
-        self._href_to_status_a = dict((meta['href'], (ident, meta))
-                                      for ident, (meta, _)
-                                      in self._ident_to_props.items())
+    @abc.abstractmethod
+    def get_new_b(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def iter_old(self):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def iter_new(self):
+        raise NotImplementedError()
 
-        self._href_to_status_b = dict((meta['href'], (ident, meta))
-                                      for ident, (_, meta)
-                                      in self._ident_to_props.items())
+    @abc.abstractmethod
+    def get_by_href_a(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_by_href_b(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def rollback(self, ident):
+        raise NotImplementedError()
+
+
+class _SqliteStatus(_StatusBase):
+    SCHEMA_VERSION = 1
+
+    def __init__(self, path):
+        self._path = path
+        self._conn = sqlite3.connect(path)
+        self._conn.row_factory = sqlite3.Row
+        self._c = None
+
+        if self._is_latest_version():
+            return
+
+        self._conn.execute('DROP TABLE IF EXISTS status')
```

BTW this doesn't silently fail, because `CREATE TABLE` will fail for existing schemas.